### PR TITLE
RSS Feed

### PR DESCRIPTION
**Pull request by johnnyramos**
Taking a first stab at getting an RSS feed generated while building the structured data feed. Let me know what you all think and if it's the right direction here.

Feedback is welcome!

# Description

This RSS feed will consist of all Security Releases from each version of the OS. `build-sofa-feed.py` will take the `feed_structure` data and sort it out by creating a new list that just consists of `SecurityReleases` and sorts them by `ReleaseDate` while splitting away the `XProtect` data from `macOS` feed into it's own.

It will then cache the json data generated into a `cache` directory for use later to differentiate from new entries that are pulled from `feed_structure`.

It then uses [feedgen](https://pypi.org/project/feedgen/) module to create the XML that will be used as the RSS feed within the `write_data_to_rss` function. RSS feed entries usually get sorted by publish dates per OS, so in this case - we use the `ReleaseDate` as the publish date.

# Changes

## build-sofa-feed.py

- Added `ProductName` and `ReleaseType` to Data Feed. This allowed me to drive the structure of the RSS feed a bit easier and less code.
- For `ReleaseType`, the default is `OS` but when it comes to `Rapid Security Response` items, it's set to `RSR_{letter}`. For example, either `RSR_a` or `RSR_c`. It does a regex on the `UpdateName` to grab it.

### load_rss_data_cache

- Loads RSS data from cache files and returns combined data.

### write_os_data_to_cache

- Writes updated OS data to the cache.
- Sorts data by ReleaseDate before writing.
- Ensures the cache directory exists and handles exceptions during file writing.

### sort_data

- Sorts data by a specified key.
- Ensures data is sorted in ascending order by default.

### diff_rss_data

- Function used to check for new entries based on `ProductVersion`
- Diffs feed results and updates cache files for each OS type.
- Identifies and writes new entries to the cache.
- Merges cache and new data entries, removes duplicates, and sorts the combined data.

### remove_dict_duplicates

- Removes duplicate dictionaries from a list.
- Uses JSON serialization to handle unhashable types.

### create_rss_json_data

- Generates feed list from the given feed structure.
- Handles XProtect data separately and mimics other entries.

### write_data_to_rss

- Writes the sorted feed data to an RSS feed file.
- Sets up primary feed information and creates individual entries.

## index.html

- Added RSS Feed icon next to the GitHub icon, linked to the single feed file created here.
- Some indentation fixes.
- Variable cleanup for data feed links.

## README.md

Updated to account for RSS Feed

## docker workflows

- Accounting for new files generated and copied over

## requirements.txt

- Adding the feedgen module

# Tests

We ran the following lines to create the new xml files along with json files. Copied them into the `v1` directory for the webserver to pick up in testing. The code also created the cache structure within `cache/` for future use.
```bash
python3 ./build-sofa-feed.py iOS
python3 ./build-sofa-feed.py macOS
python3 ./build-sofa-feed.py macOS iOS
```

Established a webserver to view changes.
```
(.venv) ➜  sofa git:(rss_feed) ✗ python3 -m http.server 80           
Serving HTTP on :: port 80 (http://[::]:80/) ...
::1 - - [12/Jun/2024 11:26:45] "GET / HTTP/1.1" 200 -
::1 - - [12/Jun/2024 11:26:45] "GET /main.css HTTP/1.1" 304 -
::1 - - [12/Jun/2024 11:26:45] "GET /images/custom_logo.png HTTP/1.1" 200 -
::1 - - [12/Jun/2024 11:26:45] "GET /config.json HTTP/1.1" 304 -
::1 - - [12/Jun/2024 11:26:45] "GET /v1/macos_data_feed.json HTTP/1.1" 200 -
::1 - - [12/Jun/2024 11:26:45] "GET /v1/timestamp.json HTTP/1.1" 304 -
::1 - - [12/Jun/2024 11:26:45] "GET /essential_links.json HTTP/1.1" 304 -
::1 - - [12/Jun/2024 11:26:45] "GET /images/Sonoma.png HTTP/1.1" 200 -
::1 - - [12/Jun/2024 11:26:45] "GET /images/SWUpdate.png HTTP/1.1" 200 -
::1 - - [12/Jun/2024 11:26:53] "GET /v1/rss_feed.xml HTTP/1.1" 200 -
```

Docker build of image/
```
➜  sofa git:(rss_feed) ✗ docker build -t sofa_jramos .
[+] Building 16.0s (21/21) FINISHED                                                                                                   docker:desktop-linux
 => [internal] load build definition from dockerfile                                                                                                  0.0s
 => => transferring dockerfile: 1.12kB                                                                                                                0.0s
 => [internal] load metadata for docker.io/library/python:3.12-alpine                                                                                 2.1s
 => [internal] load .dockerignore                                                                                                                     0.0s
 => => transferring context: 2B                                                                                                                       0.0s
 => [internal] load build context                                                                                                                     0.0s
 => => transferring context: 73.55kB                                                                                                                  0.0s
 => [ 1/16] FROM docker.io/library/python:3.12-alpine@sha256:d24ed567ee3b972478a232ceff84b0d002e18ee9f5d38234ecbffece23dfa084                         3.5s
 => => resolve docker.io/library/python:3.12-alpine@sha256:d24ed567ee3b972478a232ceff84b0d002e18ee9f5d38234ecbffece23dfa084                           0.0s
 => => sha256:820650aa2f5b8e795b04058fa617fb87ad3397f0fc2582400f9855b84f2d4401 465.47kB / 465.47kB                                                    0.8s
 => => sha256:4e0f3549117ae48d35d3cbea60bca5ef0d654723e8065274c93137c3f2b44a60 14.68MB / 14.68MB                                                      2.7s
 => => sha256:d24ed567ee3b972478a232ceff84b0d002e18ee9f5d38234ecbffece23dfa084 1.65kB / 1.65kB                                                        0.0s
 => => sha256:6714e211c290dc39908616ed0befdcc8f9097b69ee62d17f2f1833d764b93ef0 1.37kB / 1.37kB                                                        0.0s
 => => sha256:0fe205a492b75b71368b95db211dc148d50240d50752f04b34d1024084561296 6.64kB / 6.64kB                                                        0.0s
 => => sha256:94747bd812346354fd5042870b6e44e5406880a4e6b5736a9cf9c753110df11b 4.09MB / 4.09MB                                                        0.8s
 => => extracting sha256:94747bd812346354fd5042870b6e44e5406880a4e6b5736a9cf9c753110df11b                                                             0.1s
 => => extracting sha256:820650aa2f5b8e795b04058fa617fb87ad3397f0fc2582400f9855b84f2d4401                                                             0.1s
 => => sha256:a5b6895e2863b5497adb5d68fe8e5241835c9624c9a31fd2b8532b08da56b638 242B / 242B                                                            1.0s
 => => sha256:629a421706977cd903f62bb22483a0fa6cba236a94d23543e616bf5c24516467 2.74MB / 2.74MB                                                        3.3s
 => => extracting sha256:4e0f3549117ae48d35d3cbea60bca5ef0d654723e8065274c93137c3f2b44a60                                                             0.3s
 => => extracting sha256:a5b6895e2863b5497adb5d68fe8e5241835c9624c9a31fd2b8532b08da56b638                                                             0.0s
 => => extracting sha256:629a421706977cd903f62bb22483a0fa6cba236a94d23543e616bf5c24516467                                                             0.2s
 => [ 2/16] WORKDIR /app                                                                                                                              0.1s
 => [ 3/16] COPY build-sofa-feed.py /app/                                                                                                             0.0s
 => [ 4/16] COPY process_uma.py /app/                                                                                                                 0.0s
 => [ 5/16] COPY process_ipsw.py /app/                                                                                                                0.0s
 => [ 6/16] COPY config.json /app/                                                                                                                    0.0s
 => [ 7/16] COPY feed_structure_template_v1.yaml /app/                                                                                                0.0s
 => [ 8/16] COPY forked_builds.json /app/                                                                                                             0.0s
 => [ 9/16] COPY sofa-time-series.py /app/                                                                                                            0.0s
 => [10/16] COPY requirements.txt /app/                                                                                                               0.0s
 => [11/16] RUN touch time-series.csv || true                                                                                                         0.1s
 => [12/16] COPY time-series.csv /app/                                                                                                                0.0s
 => [13/16] COPY model_identifier_*.json /app/                                                                                                        0.0s
 => [14/16] RUN pip install --no-cache-dir -r requirements.txt                                                                                        9.9s
 => [15/16] COPY entrypoint.sh /app/                                                                                                                  0.0s
 => [16/16] RUN chmod +x /app/entrypoint.sh                                                                                                           0.1s
 => exporting to image                                                                                                                                0.1s
 => => exporting layers                                                                                                                               0.1s
 => => writing image sha256:ca4f664f32ada21ebfe13b6bf1a53eba8a87f4ad0eaa4b6ec94df6db62459944                                                          0.0s
 => => naming to docker.io/library/sofa_jramos
```

# Visuals

Screenshot 2024-06-08 at 12 03 40 AM

Screenshot 2024-06-12 at 11 46 03 AM
Screenshot 2024-06-12 at 11 45 50 AM
Screenshot 2024-06-12 at 11 51 20 AM
Screenshot 2024-06-12 at 11 51 55 AM

### CODE REVIEW

**Comment by headmin** [comment]
@johnnyramos As discussed the merge #80 is now live. We keep create two separate feed files based on OS type, but do so in a single run. Hopefully this is making RSS feed creation easier.

**Comment by johnnyramos** [comment]
@headmin ok, should be up to date now and tested locally for create 2 feeds in a run. Give it a whirl.

**Comment by johnnyramos** [comment]
@headmin - fixed the structure of the feed in 8f2f226.

As for reducing the details, I'm torn.
- I agree with you if we were able to link to the tab and details of the feed item. Example, `https://sofa.macadmins.io/macOS14#guid_goes_here` or similar.
- For now, we are able to provide the details in which can be digested immediately within a persons RSS app or Slack.

If we want to just keep it to the basics, I would need to submit a commit to link back to the initial page (https://sofa.macadmins.io/) for every item until we can link to specific `div`'s. I'm open to either or but just wanted to pass my opinion by you.

File: .github/workflows/run_docker_macos_workflow.yml
Comment (headmin) [inline .github/workflows/run_docker_macos_workflow.yml:1]: We need to add the XML files to git, a small change line 47 of the workflow file works here
`git add v1/*.json v1/*.xml time-series.csv `

Comment (johnnyramos) [reply .github/workflows/run_docker_macos_workflow.yml:1]: Add the cache files as well. Looks like I missed it in my last commit.

**Pull request merged**